W-10877309 Exchange create custom policy for Flex Gateway jh (#188)
mulesoft/docs-exchange: modules/ROOT/pages/asset-versions.adoc

= Asset Versions
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images/

> Each asset in Exchange is versioned. You can manage which versions are visible by deprecating a version to hide it, and delete versions if needed. All versions of an asset always have the same type.

Reviewer: ```suggestion
Each asset in Exchange is versioned. You can manage which versions are visible by deprecating a version to hide it and deleting unneeded versions. All versions of an asset have the same type.
```

> In addition to asset versions, APIs have a consumer facing API version that is shown at the top of an asset's detail screen. This API version is defined by API providers.

Reviewer: ```suggestion
APIs also have a consumer-facing API version that appears at the top of an asset's detail screen. API providers define this version.
```

Exchange asset versions follow the https://semver.org/[Semantic Versioning] model of major, minor, and patch releases. For example, if an asset is version 2.4.6, then its major version is 2.x.x, its minor version is 2.4.x, and its patch version is 2.4.6.

To select a minor version of an asset, use the menu next to the asset name at the top of the asset portal. To select a patch version of an asset, use the version list at the right.

> RAML versions have been automatically generated for all OAS assets published since xref:release-notes::exchange/anypoint-exchange-release-notes.adoc#january-2019[January 12, 2019], and OAS 2.0 versions have been automatically generated for all RAML assets published since that date.

Reviewer: ```suggestion
RAML versions are automatically generated for all OAS assets published since xref:release-notes::exchange/anypoint-exchange-release-notes.adoc#january-2019[January 12, 2019]. OAS 2.0 versions are automatically generated for all RAML assets published since that date.
```

== Asset Limits

An asset's version format must use semantic versioning. Asset fields have these length limits:

[%header,cols="2a,1a,1a"]
|===
|Field |Maximum Length |Minimum Length
|Name |256 |1
|Description |256 |0
|`versionGroup` |256 |1
|Combined `groupId`, `assetId`, and `version` |215 |NA
|===

Exchange prevents resource exhaustion attacks by limiting the number of asset versions that can be published.

> The limit is 500 assets for root organizations with trial accounts, and 100,000 (one hundred thousand) assets for other root organizations.

Reviewer: ```suggestion
Root organizations with trial accounts can have up to 500 assets and other root organizations can have up to 100,000 assets.
```

> The limit of dependencies for an asset is 100. For example, in Anypoint API Designer a RAML is limited to 100 RAML Fragment dependencies, and in Anypoint API Manager an API Group is limited to 100 REST API dependencies.

Reviewer: ```suggestion
The limit of dependencies for an asset is 100. For example, in Anypoint API Designer, a RAML is limited to 100 RAML Fragment dependencies, and in Anypoint API Manager, an API Group is limited to 100 REST API dependencies.
```

> The limit of dependencies for an asset is 100. For example, in Anypoint API Designer a RAML is limited to 100 RAML Fragment dependencies, and in Anypoint API Manager an API Group is limited to 100 REST API dependencies.

Reviewer: When you say "a RAML" should this be a "RAML file"?

When a root organization reaches 80% of its asset limit, Exchange shows a warning.

When a root organization reaches its asset limit, Exchange shows an error. Also, using the *Publish new asset* button on the home page or the *Add new version* button on the asset detail page shows an error explaining that the limit is reached and no more assets can be added.

Check the limit of your root organization with a curl command like this:

----
    curl -X GET \
      https://anypoint.mulesoft.com/exchange/api/v1/organizations/ROOT_ORGANIZATION_ID/limits \
      -H 'Authorization: bearer ANYPOINT_TOKEN'
----

Replace `ROOT_ORGANIZATION_ID` with your root organization ID, and replace `ANYPOINT_TOKEN` with the authorization token that has permissions for the root organization.

Get lists of assets with a curl command like this:

----
    curl -X POST \
      https://anypoint.mulesoft.com/graph/api/v1/graphql \
      -H 'content-type: application/json' \
      -H 'Authorization: bearer ANYPOINT_TOKEN' \
      -d '{"query":"{assets(query: { masterOrganizationId:\"ROOT_ORGANIZATION_ID\", limit: 20, offset: 10 }) {groupId assetId version}}"}'
----

> Replace `ROOT_ORGANIZATION_ID` with your root organization ID, and replace `ANYPOINT_TOKEN` with the authorization token that has permissions for the root organization. Vary the `limit` and `offset` values as needed.

Reviewer: ```suggestion
Replace `ROOT_ORGANIZATION_ID` with your root organization ID, and replace `ANYPOINT_TOKEN` with the authorization token that has permissions for the root organization. Change the `limit` and `offset` values as needed.
```
